File: build-portable.mjs
/**
 * build-portable.mjs
 * Builds a portable IDEpro ZIP from:
 *   - src/           (patched source)
 *   - Electron binary from installed Antigravity IDE
 *
 * Output: dist/IDEpro-Portable.zip
 *
 * Steps:
 *   1. Pack src/ → dist/resources/app.asar  (using @electron/asar)
 *   2. Copy Electron binary + DLLs → dist/
 *   3. Rename electron binary → IDEpro.exe
 *   4. Zip dist/ → IDEpro-Portable.zip
 */

import { execSync } from "node:child_process";
import { cpSync, mkdirSync, rmSync, existsSync, copyFileSync, readdirSync, statSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const SRC_DIR          = path.join(__dirname, "package", "resources", "app");
const DIST_DIR          = path.join(__dirname, "dist", "IDEpro");
const ELECTRON_DIR     = "C:\\Users\\Open Pc\\AppData\\Local\\Programs\\Antigravity IDE";
const OUTPUT_ZIP       = path.join(__dirname, "dist", "IDEpro-Portable.zip");

// ─── Step 0: Clean dist ───────────────────────────────────────────────────────
console.log("🧹 Cleaning dist...");
if (existsSync(DIST_DIR)) rmSync(DIST_DIR, { recursive: true });
mkdirSync(path.join(DIST_DIR, "resources"), { recursive: true });

// ─── Step 1: Pack app.asar ────────────────────────────────────────────────────
console.log("📦 Packing app.asar...");
const asarOut = path.join(DIST_DIR, "resources", "app.asar");
execSync(`npx --yes @electron/asar pack "${SRC_DIR}" "${asarOut}"`, {
    stdio: "inherit",
    cwd: __dirname,
});
console.log("✅ app.asar packed");

// ─── Step 2: Copy app.asar.unpacked (native modules) ─────────────────────────
const unpackedSrc = path.join(ELECTRON_DIR, "resources", "app.asar.unpacked");
if (existsSync(unpackedSrc)) {
    console.log("📋 Copying app.asar.unpacked...");
    cpSync(unpackedSrc, path.join(DIST_DIR, "resources", "app.asar.unpacked"), { recursive: true });
}

// ─── Step 3: Copy Electron binary + all DLLs/files ────────────────────────────
console.log("🔧 Copying Electron runtime...");

// Copy everything EXCEPT the original exe and uninstaller
const skipFiles = new Set(["Antigravity IDE.exe", "unins000.exe", "unins000.dat"]);
const entries = readdirSync(ELECTRON_DIR);

for (const entry of entries) {
    if (skipFiles.has(entry)) continue;
    const src = path.join(ELECTRON_DIR, entry);
    const dst = path.join(DIST_DIR, entry);
    const stat = statSync(src);
    if (stat.isDirectory()) {
        cpSync(src, dst, { recursive: true });
    } else {
        copyFileSync(src, dst);
    }
}

// ─── Step 4: Rename/copy electron EXE → IDEpro.exe ───────────────────
console.log("📝 Renaming EXE to IDEpro.exe...");
copyFileSync(
    path.join(ELECTRON_DIR, "Antigravity IDE.exe"),
    path.join(DIST_DIR, "IDEpro.exe")
);

// ─── Step 5: Create ZIP ────────────────────────────────────────────────────────
console.log("🗜️  Creating ZIP...");
if (existsSync(OUTPUT_ZIP)) rmSync(OUTPUT_ZIP);

// Use PowerShell Compress-Archive
execSync(
    `powershell -Command "Compress-Archive -Path '${DIST_DIR}\\*' -DestinationPath '${OUTPUT_ZIP}' -Force"`,
    { stdio: "inherit" }
);

console.log("\n✅ Done!");
console.log(`📦 Output: ${OUTPUT_ZIP}`);
console.log(`📁 Portable folder: ${DIST_DIR}`);
console.log(`\n💡 To run directly: ${DIST_DIR}\\IDEpro.exe`);

File: build_portable.py
import os
import sys
import shutil
import subprocess

# Set console encoding to UTF-8 on Windows to avoid UnicodeEncodeError for emojis
if sys.platform.startswith("win"):
    try:
        sys.stdout.reconfigure(encoding="utf-8")
        sys.stderr.reconfigure(encoding="utf-8")
    except AttributeError:
        pass

def main():
    root_dir = os.path.dirname(os.path.abspath(__file__))
    src_dir = os.path.join(root_dir, "package", "resources", "app")
    dist_parent = os.path.join(root_dir, "dist")
    dist_dir = os.path.join(dist_parent, "IDEpro")
    electron_dir = r"C:\Users\Open Pc\AppData\Local\Programs\Antigravity IDE"
    output_zip = os.path.join(dist_parent, "IDEpro-Portable.zip")

    # Step 0: Clean dist
    print("🧹 Cleaning dist...")
    if os.path.exists(dist_dir):
        shutil.rmtree(dist_dir)
    os.makedirs(os.path.join(dist_dir, "resources"), exist_ok=True)

    # Step 1: Pack app.asar
    print("📦 Packing app.asar...")
    asar_out = os.path.join(dist_dir, "resources", "app.asar")
    try:
        subprocess.run(
            ["npx", "--yes", "@electron/asar", "pack", src_dir, asar_out],
            check=True,
            shell=True,
            cwd=root_dir
        )
        print("✅ app.asar packed")
    except Exception as e:
        print(f"❌ Failed to pack app.asar: {e}", file=sys.stderr)
        sys.exit(1)

    # Step 2: Copy app.asar.unpacked (native modules)
    unpacked_src = os.path.join(electron_dir, "resources", "app.asar.unpacked")
    if os.path.exists(unpacked_src):
        print("📋 Copying app.asar.unpacked...")
        shutil.copytree(unpacked_src, os.path.join(dist_dir, "resources", "app.asar.unpacked"))

    # Step 3: Copy Electron runtime
    print("🔧 Copying Electron runtime...")
    skip_files = {"Antigravity IDE.exe", "unins000.exe", "unins000.dat"}
    for entry in os.listdir(electron_dir):
        if entry in skip_files:
            continue
        src = os.path.join(electron_dir, entry)
        dst = os.path.join(dist_dir, entry)
        if os.path.isdir(src):
            shutil.copytree(src, dst, dirs_exist_ok=True)
        else:
            shutil.copy2(src, dst)

    # Step 4: Rename/copy electron EXE -> IDEpro.exe
    print("📝 Renaming EXE to IDEpro.exe...")
    shutil.copy2(
        os.path.join(electron_dir, "Antigravity IDE.exe"),
        os.path.join(dist_dir, "IDEpro.exe")
    )

    # Step 5: Create ZIP
    print("🗜️ Creating ZIP...")
    if os.path.exists(output_zip):
        os.remove(output_zip)

    # Create zip from dist_dir content (but exclude dist_dir base folder inside zip, just pack its children)
    zip_base = os.path.splitext(output_zip)[0]
    shutil.make_archive(zip_base, "zip", dist_dir)

    print("\n✅ Done!")
    print(f"📦 Output: {output_zip}")
    print(f"📁 Portable folder: {dist_dir}")
    print(f"\n💡 To run directly: {dist_dir}\\IDEpro.exe")

if __name__ == "__main__":
    main()
